staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go

```diff
@@ -37,6 +38,42 @@ func getColumnsForVersion(crd *apiextensionsv1.CustomResourceDefinition, version
 	return nil, fmt.Errorf("version %s not found in apiextensionsv1.CustomResourceDefinition: %v", version, crd.Name)
 }
 
+// getScaleColumnsForVersion returns 2 columns for the desired and actual number of replicas.
+func getScaleColumnsForVersion(crd *apiextensionsv1.CustomResourceDefinition, version string) ([]apiextensionsv1.CustomResourceColumnDefinition, error) {
+	for _, v := range crd.Spec.Versions {
+		if version != v.Name {
+			continue
+		}
+		var cols []apiextensionsv1.CustomResourceColumnDefinition
+		if v.Subresources != nil && v.Subresources.Scale != nil {
+			if v.Subresources.Scale.SpecReplicasPath != "" {
+				cols = append(cols, apiextensionsv1.CustomResourceColumnDefinition{
+					Name:        "Desired",
+					Type:        "integer",
+					Description: "Number of desired replicas",
+					JSONPath:    ".spec.replicas",
+				})
+			}
+			if v.Subresources.Scale.StatusReplicasPath != "" {
+				cols = append(cols, apiextensionsv1.CustomResourceColumnDefinition{
+					Name:        "Available",
+					Type:        "integer",
+					Description: "Number of actual replicas",
+					JSONPath:    ".status.replicas",
+				})
+			}
+		}
+		cols = append(cols, apiextensionsv1.CustomResourceColumnDefinition{
+			Name:        "Age",
+			Type:        "date",
+			Description: swaggerMetadataDescriptions["creationTimestamp"],
+			JSONPath:    ".metadata.creationTimestamp",
+		})
+		return cols, nil
+	}
+	return nil, fmt.Errorf("version %s not found in apiextensionsv1.CustomResourceDefinition: %v", version, crd.Name)
```

Comment:
Maybe I haven't seen the place where this happens, but it seems like we should pre-check this and not support the operation at all if this is the case?

Reply:
@lavalamp you're right. I think the error was added to stay consistent with pre-existining logic:
https://github.com/kubernetes/kubernetes/blob/95e30f66c300c76ce21c0ca0e8bc4bf4a45e028f/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/helpers.go#L31

However, one curious thing is that the error will never be reached since these functions are called in a loop over `crd.spec.versions`:
https://github.com/kubernetes/kubernetes/blob/95e30f66c300c76ce21c0ca0e8bc4bf4a45e028f/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L709

Let me add a comment for now, and I can help cleanup the error parts as a follow up.

Reply:
OK

Reply:
(ordinarily I'd ask to refactor until there's no way it can go wrong, but this PR has been in flight long enough and the readability gain from this is minimal -- of course the iterated version of this is why codebases get harder and harder to understand over time, but that doesn't mean it's the wrong tradeoff to make now)